File: client/ui/frontend/src/lib/compat.ts
import { Compat } from "@bindings/services";

let cached: boolean | null = null;

/**
 * isDaemonCompatible probes whether the running daemon implements the WailsUIReady
 * RPC. A false result means the daemon predates this UI (Unimplemented) and is too
 * old to drive it. The Go side returns an error instead when the daemon is simply
 * unreachable, so a throw here is NOT an outdated daemon — treat it as "unknown"
 * and let the normal connection flow report it.
 *
 * The result is cached for the session: daemon identity does not change without a
 * UI restart, and a freshly started daemon is reachable again under the same socket.
 */
export async function isDaemonCompatible(): Promise<boolean> {
    if (cached !== null) return cached;
    cached = await Compat.DaemonReady();
    return cached;
}

File: client/ui/services/compat.go
//go:build !android && !ios && !freebsd && !js

package services

import (
	"context"

	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"

	"github.com/netbirdio/netbird/client/proto"
)

// Compat answers whether the running daemon is new enough to drive this UI.
type Compat struct {
	conn DaemonConn
}

func NewCompat(conn DaemonConn) *Compat {
	return &Compat{conn: conn}
}

// DaemonReady probes the WailsUIReady RPC once. A true result means the daemon
// implements it and is compatible. An Unimplemented response means the daemon
// predates this UI and is too old; the caller should surface an upgrade prompt.
// Any other error (daemon not running, transport failure) is returned so the
// frontend can tell "outdated" apart from "not reachable".
func (c *Compat) DaemonReady(ctx context.Context) (bool, error) {
	client, err := c.conn.Client()
	if err != nil {
		return false, err
	}
	if _, err := client.WailsUIReady(ctx, &proto.WailsUIReadyRequest{}); err != nil {
		if st, ok := status.FromError(err); ok && st.Code() == codes.Unimplemented {
			return false, nil
		}
		return false, err
	}
	return true, nil
}
